### ISSUE TITLE
Ganga fails to lock files on NFSv4 mount

### ISSUE THREAD

**Issue by stdietrich** [ganga-devs/ganga#959]
Hello,

I am currently debugging an issue with a user, who tries to run Ganga v602r3 on our batch farm BIRD/NAF at DESY.
Ganga tries to store the output file on an NFSv4 exported GPFS filesystem, however it fails with the following error message:

```
ERROR Error: [Errno 9] Bad file descriptor
ERROR A severe error occurred in the Repository 'prep.metadata':

Unable to launch due to underlying filesystem not working with unix
locks.Please try launching again with
[Configuration]lockingStrategy=FIXED to start Ganga without multiple
session support.
ERROR If you believe the problem has been solved, type "reactivate()"
to re-enable
INFO Ganga is now attempting to shut down all running processes
accessing the repository in a clean manner
INFO ... Please be patient!
ERROR Unable to disable Internal services, they may have already been
disabled!
```

I have traced this error back to the [dry_run_unix_locks](https://github.com/ganga-devs/ganga/blob/6c4a7cfb75a138b14e7f8c364faee3c27413e116/python/Ganga/Core/GangaRepository/SessionLock.py#L258) function in SessionLock.py.
The problematic part seems to be the following line:

`with open(test_file) as f_lock:`

The file is opened read-only, however this behaves different on NFSv3 and NFSv4.
On NFSv3 exports, opening the file as RO is working, while on NFSv4 the Bad File Descriptor error message is logged.
According to a post on the [NFS mailing list](http://www.spinics.net/lists/linux-nfs/msg18594.html), this is an expected behavior.

If the lock file is opened with Read/Write, it works on NFSv4.
Please let me know, if you need more information.

Regards,
Stefan

**Comment by egede** [comment]
@stdietrich This is interesting. We have had problems on systems before with the `f_lock` but they have usually been related to badly configured systems. I have not come across this discussion of if it depends on how the file is opened before. Can you check on the system in question if the `f_lock` actually works if you open the file in read/write mode? As this particular file locking has caused trouble in other places and that there are performance issues related to it, we have to be a bit careful about just changing it.

**Comment by stdietrich** [comment]
What do you mean, if Ganga works with f_lock in read/write or just the function?
The latter I already verified, because I created a sample script based on that particular function.
Running `with open(test_file, 'w') as f_lock:` works and the bad file descriptor error is gone.

For the first part, at least there I have to forward this to the user and let him try again.

**Comment by egede** [comment]
@stdietrich I just meant in a sample script, so nice to hear this is already done. I will create a pull request with the change and then maybe you can get the user to try that version of `Ganga`.

**Comment by egede** [comment]
@stdietrich You can try to use the pull request #960 and report if that is working.

**Comment by egede** [comment]
Fixed by #960
